File: leetcode笔记/栈/155. easy.最小栈/min_stack.cpp
class MinStack {
	
public:
    MinStack() {
    }
	
    // 入栈
    void push(int x) {
        stk.push(x);
        if (!minstk.empty()) {
            // 比 minstk 栈的栈顶小才入栈
            // 这样 minstk 栈的栈顶一定是最小的 因为比栈顶大的不会进这个栈
            minstk.push( min(minstk.top(), x) ); 
        }   
        else
            minstk.push(x);
    }
    
    // 出栈
    void pop() {
        stk.pop();
        minstk.pop();
    }
    
    // 获取栈顶
    int top() {
        return stk.top();
    }
    
    // 返回栈最小元素
    int getMin() {
        return minstk.top();
    }
	
	// 借助两个栈
    stack<int> stk, minstk;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: leetcode笔记/栈/155. easy.最小栈/min_stack.py
class MinStack(object):
    
    def __init__(self):
        self.stk = []
        self.minstk = []
        
    # 入栈
    def push(self, x):
        self.stk.append(x)
        # minstk的栈顶元素一定要为最小，所以要入栈的值要比minstk栈的栈顶元素小才入栈
        if len(self.minstk) == 0 or x <= self.minstk[-1]:
            self.minstk.append(x)
        
    # 出栈
    def pop(self):
        tmp = self.stk.pop()
        # 有可能之前入栈的元素中有比最小栈栈顶大的，只入了stk栈而没入minstk栈
        if tmp == self.minstk[-1]:
            self.minstk.pop()
        
    # 获取栈顶元素
    def top(self):
        return self.stk[-1]

    # 获取栈内最小元素
    def getMin(self):
        return self.minstk[-1]